erts/emulator/beam/erl_bif_persistent.c

```diff
@@ -0,0 +1,816 @@
+        hp += 3;
+        res = CONS(hp, tup, res);
+        hp += 2;
+    }
+    erts_free(ERTS_ALC_T_TMP, copy_data);
+    return res;
+}
+
+static BIF_RETTYPE
+persistent_term_info_trap(BIF_ALIST_1)
+{
+    TrapData* trap_data = (TrapData *) BIF_ARG_1;
+    Eterm res;
+
+    ASSERT(is_big(BIF_ARG_1));
+    trap_data = (TrapData *) boxed_val(BIF_ARG_1);
+    res = do_info(BIF_P, trap_data);
+    if (trap_data->remaining > 0) {
+        ASSERT(res == am_ok);
+        BUMP_ALL_REDS(BIF_P);
+        BIF_TRAP1(&persistent_term_info_export, BIF_P, BIF_ARG_1);
+    } else {
+        /*
+         * Decrement ref count (and possibly delete the hash table
+         * and associated literal area).
+         */
+        dec_table_refc(BIF_P, trap_data->table);
+        ASSERT(is_map(res));
+        BIF_RET(res);
+    }
+}
+
+#define DECL_AM(S) Eterm AM_ ## S = am_atom_put(#S, sizeof(#S) - 1)
+#define NUM_INFO_MAP_KEYS 2
+
+static Eterm
+do_info(Process* c_p, TrapData* trap_data)
+{
+    HashTable* hash_table;
+    Uint remaining;
+    Uint idx;
+    Uint max_iter;
+
+    hash_table = trap_data->table;
+    idx = trap_data->idx;
+#if defined(DEBUG) || defined(VALGRIND)
+    max_iter = 50;
+#else
+    max_iter = c_p->fcalls;
+#endif
+    remaining = trap_data->remaining < max_iter ? trap_data->remaining : max_iter;
+    trap_data->remaining -= remaining;
+    while (remaining != 0) {
+        if (is_boxed(hash_table->term[idx])) {
+            ErtsLiteralArea* area;
+            area = term_to_area(hash_table->term[idx]);
+            trap_data->memory += sizeof(ErtsLiteralArea) +
+                sizeof(Eterm) * (area->end - area->start - 1);
+            remaining--;
+        }
+        idx++;
+    }
+    trap_data->idx = idx;
+    if (trap_data->remaining > 0) {
+        return am_ok;           /* Dummy return value */
+    } else {
+        Eterm* hp;
+        Eterm count_term;
+        Eterm memory_term;
+        Eterm res;
+        Uint memory;
+        Uint hsz = MAP_SZ(2);
+
+        memory = sizeof(HashTable) + (trap_data->table->allocated-1) *
+            sizeof(Eterm) + trap_data->memory;
+        (void) erts_bld_uint(NULL, &hsz, hash_table->num_entries);
+        (void) erts_bld_uint(NULL, &hsz, memory);
+        hp = HAlloc(c_p, hsz);
+	count_term = erts_bld_uint(&hp, NULL, hash_table->num_entries);
+	memory_term = erts_bld_uint(&hp, NULL, memory);
+        res = MAP2(hp, am_count, count_term, am_memory, memory_term);
+        return res;
+    }
+}
+
+#undef NUM_INFO_MAP_KEYS
+#undef DECL_AM
+
+static Uint
+lookup(HashTable* hash_table, Eterm key)
+{
+    Uint mask = hash_table->mask;
+    Eterm* table = hash_table->term;
+    Uint32 idx = make_internal_hash(key, 0);
+    Eterm term;
+
+    do {
+        idx++;
+        term = table[idx & mask];
+    } while (is_boxed(term) && !EQ(key, (tuple_val(term))[1]));
+    return idx & mask;
+}
+
+static HashTable*
+copy_table(HashTable* old_table, Uint new_size, int rehash)
```

Comment:
How large do we expect this hash table to get? Do we need to trap while copying it?

Reply:
The recommendation in the documentation is to limit the total number of persistent terms. Therefore, I have assumed that the table is always small enough so that we don't have to trap.